src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.RenameNamespaceCodeAction.cs

```diff
@@ -0,0 +1,295 @@
+
+                    // Ignore other documents with identical file path as triggering document (i.e. other TFM in a MTFM project)
+                    refLocationsInCurrentDocument = refLocationsInCurrentDocument.Concat(refLocationsForSymbol.Where(loc => loc.Document.Id == document.Id));
+                    refLocationsInOtherFilePath = refLocationsInOtherFilePath.Concat(refLocationsForSymbol.Where(loc => !PathUtilities.PathsEqual(loc.Document.FilePath, document.FilePath)));
+                }   
+
+                // Transform current document(s) first
+                solution = (await FixDeclarationDocumentAsync(document, refLocationsInCurrentDocument, cancellationToken).ConfigureAwait(false)).Project.Solution;
+
+                // Then fix all referencing documents.
+                var oldImport = _service.CreateUsingDirective(OldNamespaceParts);
+                var newImport = _service.CreateUsingDirective(NewNamespaceParts);
+
+                var refLocationGroups = refLocationsInOtherFilePath.GroupBy(loc => loc.Document.Id);
+                foreach (var refInOneDocument in refLocationGroups)
+                {
+                    var refDocument = solution.GetDocument(refInOneDocument.Key);
+                    solution = (await FixReferencingDocumentAsync(refDocument, oldImport, newImport, refInOneDocument, cancellationToken).ConfigureAwait(false)).Project.Solution;
+                }
+
+                return ImmutableArray.Create<CodeActionOperation>(new ApplyChangesOperation(solution));
+            }            
+
+            private async Task<Document> FixDeclarationDocumentAsync(Document document, IEnumerable<ReferenceLocation> refLocations, CancellationToken cancellationToken)
+            {
+                Debug.Assert(!NewNamespaceParts.IsDefault);
+
+                // 1. Reduce all references.
+                // 2. Add usings for containing namespaces, in case we have references relying on old namespace declaration for resolution. 
+                // 3. Change namespace declaration.
+                // 4. Remove unnecessary usings.              
+
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+
+                ImmutableHashSet<SyntaxNode> containers;
+                (document, containers) = await FixReferences(document, addImportService, refLocations, namespaceParts: default, cancellationToken);
+                var root = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+
+                var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                document = await Simplifier.ReduceAsync(document, optionSet, cancellationToken).ConfigureAwait(false);                
+
+                // Create import for all levels of old namespace and add them to the document (if it's not global namespace) 
```

Comment:
this doesn't explain why.  comments taht just restate the code aren't helpful.  they need to explain *why* they're doing what they're doing.